lib/AuthStateManager.ts

```diff
@@ -0,0 +1,148 @@
+  _pending: { 
+    updateAuthStatePromise: typeof PCancelable;
+    canceledTimes: number; 
+  };
+  _authState: AuthState;
+
+  constructor(sdk: OktaAuth) {
+    if (!sdk.emitter) {
+      throw new AuthSdkError('Emitter should be initialized before AuthStateManager');
+    }
+
+    this._sdk = sdk;
+    this._pending = { ...DEFAULT_PENDING };
+    this._authState = { ...DEFAULT_AUTH_STATE };
+
+    // Listen on tokenManager events to start updateState process
+    // "added" event is emitted in both add and renew process, it will be canceled by the followed "renewed" event
+    sdk.tokenManager.on('added', (key, token) => {
+      this.updateAuthState({ event: 'added', key, token });
+    });
+    sdk.tokenManager.on('renewed', (key, token) => {
+      this.updateAuthState({ event: 'renewed', key, token });
+    });
+    sdk.tokenManager.on('removed', (key, token) => {
+      this.updateAuthState({ event: 'removed', key, token });
+    });
+  }
+
+  getAuthState(): AuthState {
+    return this._authState;
+  }
+
+  updateAuthState({ event, key, token }: UpdateAuthStateOptions = {}): void {
+    const { isAuthenticated, devMode } = this._sdk.options;
+
+    const logger = (status) => {
+      console.group(`OKTA-AUTH-JS:updateAuthState: Event:${event} Status:${status}`);
+      console.log(key, token);
+      console.log('Current authState', this._authState);
+      console.groupEnd();
+    };
+
+    const emitAuthStateChange = (authState) => {
+      this._authState = authState;
+      // emit new authState object
+      this._sdk.emitter.emit(EVENT_AUTH_STATE_CHANGE, { ...authState });
+      devMode && logger('emitted');
+    };
+
+    if (this._pending.updateAuthStatePromise) {
+      if (this._pending.canceledTimes >= MAX_PROMISE_CANCEL_TIMES) {
+        // stop canceling then starting a new promise
+        // let existing promise finish to prevent running into loops
+        return;
+      } else {
+        this._pending.updateAuthStatePromise.cancel();
+      }
+    }
+
+    /* eslint-disable complexity */
+    const cancelablePromise = new PCancelable((resolve, _, onCancel) => {
+      onCancel.shouldReject = false;
+      onCancel(() => {
+        this._pending.updateAuthStatePromise = null;
+        this._pending.canceledTimes = this._pending.canceledTimes + 1;
+        devMode && logger('canceled');
+      });
+
+      const emitAndResolve = (authState) => {
+        if (cancelablePromise.isCanceled) {
+          resolve();
+          return;
+        }
+        // emit event and clear states
+        emitAuthStateChange(authState); 
+        this._pending = { ...DEFAULT_PENDING };
+        resolve();
+      };
+
+      return this._sdk.tokenManager._getTokens()
+        .then(({ accessToken, idToken }) => {
+          if (cancelablePromise.isCanceled) {
+            resolve();
+            return;
+          }
+
+          if (accessToken && this._sdk.tokenManager.hasExpired(accessToken)) {
+            accessToken = null;
```

Comment:
if options.autoRenew is true, shouldn't we attempt to renew the token here? If an autoRenew is pending we will see the UI show authorized=false briefly before the renew completes and then the app is authorized.

Reply:
I think `AuthStateManager` should only read (by `get`) tokens from tokenManager, but not write (by `renew`) to it. Also, the tokenManager should be able to handle tokens state properly and emit events to AuthStateManager. 
In this scenario, the authStateManager should expect to get a new event from tokenManager, then evaluate based on it.

Reply:
It will get a new event and re-evaluate, but the app will flash from not authenticated to authenticated. This is what "isPending" is supposed to avoid. If the initial auto renew logic has not run, then we should be in a pending state until it completes.

Reply:
I think the scenario you described above can be an issue if tokens are renewed one by one, but in the current implementation, two tokens are renewed together, the cancelable promise is able to prevent to false-positive case.

Reply:
In a test case, you could mock the `renew` method to have a long timeout, or maybe not that long - 500ms.
How long does it take `updateAuthState` to run? It will be able to read the tokens, see that they have expired and emit a new state `isAuthenticated: false` state to all listeners, easily within 500ms. When the renew completes `updateAuthState` will be called again. The app will render 3 times:
1: loading (isPending: true)
2: unauthenticated, display a login dialog (isAuthenticated: false) 
(for 1/2 second)
3: authenticated: (isAuthenticated: true)

If there was a way for the initial token renew process to tap into the "isPending" concept, it could prevent the pending state from becoming true until it has completed.